interactable scale theme now relative to gameobject initial scale

gilbdev opened the pull request:
## Overview
Make scale theme consistent with other themes by having a scale relative to initial object scale and not absolute.

## Changes
- Fixes: #6459


## Verification
> This optional section is a place where you can detail the specific type of verification 
> you want from reviewers. For example, if you want reviewers to checkout the PR locally
> and validate the functionality of specific scenarios, provide instructions
> on the specific scenarios and what you want verified.
>
> If there are specific areas of concern or question feel free to highlight them here so
> that reviewers can watch out for those issues.
>
> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

keveleigh commented:
/azp run

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableScaleTheme.cs:
```diff
@@ -59,7 +61,7 @@ public override ThemePropertyValue GetProperty(ThemeStateProperty property)
         public override void SetValue(ThemeStateProperty property, int index, float percentage)
         {
-            hostTransform.localScale = Vector3.Lerp(property.StartValue.Vector3, property.Values[index].Vector3, percentage);
+            hostTransform.localScale = Vector3.Lerp(property.StartValue.Vector3, Vector3.Scale(startScale, property.Values[index].Vector3), percentage);
```
I wouldn't call it startScale. It's very confusing with property.StartValue.Vector3. originalScale is a bit more deliberate. 

Also RotationTheme does not take into account original rotation either. 

This change will technically "break" people who were expecting the scale value to be direct vs multiplicative....but preserving the original information is more of what people would generally want I imagine. Especially if we eventually get to the work to "reset" themes allowing all changes done to be reverted. 

Curious to hear other thoughts on this? @Julenka, @CDiaz-MS , @killerantz  ?

gilbdev replied:
naming startScale : i just copy/paste code from `ScaleOffsetColorTheme` ...
i can change both to originalScale

rotation : good point, it should be all or none

breaking : sure, it can have an impact for people that use the scale theme, on object that do not have a scale at (1,1,1). depending on the change, it will need to appear in the version updating guide.

=> my point is that currently there are some theme that behave as relative changes, and other as absolute, and this need to be changed somehow to be more uniform and consistent.
i vote for the 'relative' way ;)

Troy-Ferrell replied:
I think I agree. I think relative is better...but not sure what others would expect. More importantly though, as you said, the themes should be consistent

keveleigh commented:
/azp run

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableRotationTheme.cs:
```diff
@@ -12,7 +12,8 @@ namespace Microsoft.MixedReality.Toolkit.UI
     {
-        private Transform hostTransform;
+        protected Vector3 originalRotation;
+		protected Transform hostTransform;
```
minor style comment to keep code consistent: please indent so that all members have same indentation level

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableRotationTheme.cs:
```diff
@@ -46,6 +47,7 @@ public override void Init(GameObject host, ThemeDefinition settings)
             base.Init(host, settings);
 
             hostTransform = Host.transform;
+			originalRotation = hostTransform.localEulerAngles;
```
the indentation on this seems off as well. Please update so that spacing is consistent.

Troy-Ferrell commented:
For any new themes, Don’t add Interactive or Interactabke to the front of the class name. We actually want to get away from that. We just can’t change it because it’ll break people. 

A ScaleMultiplierTheme or RelativeScaleTheme works or even ScaleTheme

gilbdev commented:
i will go with the RelativeScaleTheme and RelativeRotationTheme.
you can still tag the old themes as deprecated in futures versions to warn people that the relative themes are more suitables if you don't want to keep all themes (this will be a little smoother change than brekaing change this way i guess)

gilbdev commented:
wait.. i could also add a bool "Custom Property" 'Relative'. this will preserve a unique theme and be backward compatible (making relative false by default)

gilbdev commented:
should i also add the 'Relative' custom property in the ScaleOffsetColorTheme ?
This one would have its default value to true to be backward compatible.

Troy-Ferrell commented:
@gilbdev, Yes a property will work as well. I would say don't add it for ScaleOffsetColorTheme. That theme is already doing things relative for both position and scale. 

Look at `InteractableGrabScaleTheme` for example of how to use the `CustomProperties`

Troy-Ferrell commented:
/azp run

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/InteractableRotationTheme.cs:
```diff
@@ -36,7 +37,15 @@ public override ThemeDefinition GetDefaultThemeDefinition()
+                {
+                    new ThemeProperty()
+                    {
+                        Name = "Relative Rotation",
```
The relative scale is a nice way to add backward compatibility, thank you for this. It would be good to document the meaning of "relative rotation" either in the interactable docs or if possible in a tooltip.

gilbdev replied:
i added the possibility to add a tooltip to custom properties.

julenka replied:
Thank you!

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/ThemeEngines/ScaleOffsetColorTheme.cs:
```diff
@@ -67,8 +67,8 @@ public override void Init(GameObject host, ThemeDefinition settings)
             hostTransform = Host.transform;
-            startPosition = hostTransform.localPosition;
-            startScale = hostTransform.localScale;
+            originalPosition = hostTransform.localPosition;
```
ScaleOffsetColor theme is used quite a often in MRTK, it would be nice to add the "relative" property here as well.

gilbdev replied:
@Troy-Ferrell said it wasn't necessary on the ScaleOffsetColor :p
i can still add it but it would default to 'true'

Troy-Ferrell replied:
I don't think it's necessary because both Scale and Offset are purely relative in this theme and I was trying to remove obstacles to get the rest of the PR in. Feel free to add it though

julenka commented:
It looks like quite a few tests failed, you may want to merge the latest code and then run tests locally to make sure the changes pass tests. See [here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/UnitTests.html) for how to run tests.

gilbdev commented:
tests failed because theme created before did not have the custom property.

when forcing a serialization, the custom property is not added :/
@Troy-Ferrell i could use some help here, usually forcing a serialization should create fileds that exists in code but not in asset, and it does not seem to work this way for customproperties.

i'm adding a contextual menu option to reserialize only selection (right click on some assets or folder, Mixed Reality Toolkit > Reseriliaze Selection)
doing this on the "Theme" folder would jsut add the 'Tooltip' field i added in the previous commit.

gilbdev commented:
there was no option to test custom properties in VisualThemeTests.cs, so i added it in the generic functions and added 2 tests for relative scale and rotation.

imo, using custom properties is a bit tricky currently, code only retrieves custom property values from an array index, so it's easily breakable at some point. ANd current theme can't be easily extended with additional custom properties like here.

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/VisualThemeTests.cs:
```diff
@@ -125,6 +125,41 @@ public IEnumerator TestRotationTheme()
         }
 
+        [UnityTest]
+        public IEnumerator TestRelativeRotationTheme()
```
nit: add comment about test

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/VisualThemeTests.cs:
```diff
@@ -145,6 +180,42 @@ public IEnumerator TestScaleTheme()
+
+            Vector3 expectedState0 = new Vector3(hostInitialScale.x * state0.x, hostInitialScale.y * state0.y, hostInitialScale.z * state0.z);
+            Vector3 expectedState1 = new Vector3(hostInitialScale.x * state1.x, hostInitialScale.y * state1.y, hostInitialScale.z * state1.z);
+
```
nit: extra space

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/VisualThemeTests.cs:
```diff
@@ -145,6 +180,42 @@ public IEnumerator TestScaleTheme()
         }
 
+        [UnityTest]
+        public IEnumerator TestRelativeScaleTheme()
```
nit: add comment about test

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/VisualThemeTests.cs:
```diff
@@ -145,6 +180,42 @@ public IEnumerator TestScaleTheme()
+            Vector3 state0 = Vector3.one * 4.0f;
+            Vector3 state1 = Vector3.one;
+
+            Vector3 expectedState0 = new Vector3(hostInitialScale.x * state0.x, hostInitialScale.y * state0.y, hostInitialScale.z * state0.z);
```
Use Vector3.Scale instead? Cleaner and same result

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/VisualThemeTests.cs:
```diff
@@ -125,6 +125,41 @@ public IEnumerator TestRotationTheme()
+            Vector3 state0 = Quaternion.LookRotation(Vector3.up).eulerAngles;
+            Vector3 state1 = Quaternion.LookRotation(Vector3.down).eulerAngles;
+
+            Vector3 expectedState0 = new Vector3(hostInitialRotation.x + state0.x, hostInitialRotation.y + state0.y, hostInitialRotation.z + state0.z);
```
same here
`Vector3 expectedState0 = hostInitialRotation + state0;`

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/Core/ThemeProperty.cs:
```diff
@@ -12,6 +12,7 @@ namespace Microsoft.MixedReality.Toolkit.UI
     public class ThemeProperty
     {
         public string Name;
+        public string Tooltip;
```
while here can you add comments on each public field?

gilbdev commented:
i've gone with the "ISerializationCallbackReceiver" way, which enables to got a callback before the asset got serialized.
this gets pretty handy here.
Theme would still require a reserialization though (apparently just focusing the theme is enough), so i still have to catch the case if it happens in runtime to hint with a logwarning.

will do it tomorrow(or any other suggestion). 
out of the office from now ;)

Troy-Ferrell commented:
> i've gone with the "ISerializationCallbackReceiver" way, which enables to got a callback before the asset got serialized.
> this gets pretty handy here.
> Theme would still require a reserialization though (apparently just focusing the theme is enough), so i still have to catch the case if it happens in runtime to hint with a logwarning.
> 
> will do it tomorrow(or any other suggestion).
> out of the office from now ;)

ISerializationCallbackReceiver works too. Actually yes we still need the runtime checks. More importantly, it should check both state properties and custom properties. I can assemble a visual theme myself and not use the inspector, thus I could supply 100 custom properties if I desired to a theme engine. We should have a check that the number of properties equals expected. 

I mean better yet, we need better guards against the property access for when the theme engine is using the properties...but I think that will feature creep this PR too far

gilbdev commented:
/azp run

keveleigh commented:
/azp run

gilbdev commented:
unit tests passed here, both in the branch and when pulling mrtk_developement locally ...
what can i do ?
tried several times, all tests, or interactable tests, or only 1 test ..

keveleigh commented:
@gilbdev Tests passed this time!

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/Core/ThemeDefinition.cs:
```diff
@@ -136,5 +136,33 @@ public Easing Easing
+
+            if (defaultDefinition.CustomProperties.Count > CustomProperties.Count)
+            {
+                //Debug.LogWarning($"{Name} Theme has inconsistent custom properties, consider forcing serialization. Fixing now by adding default custom properties.");
```
commented code

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/VisualThemes/Core/ThemeDefinition.cs:
```diff
@@ -136,5 +136,33 @@ public Easing Easing
             return null;
         }
+
+        void ISerializationCallbackReceiver.OnBeforeSerialize()
```
add public specifier

gilbdev replied:
_The public keyword is not allowed on an explicit interface declaration_

julenka commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/VisualThemeTests.cs:
```diff
@@ -125,6 +125,44 @@ public IEnumerator TestRotationTheme()
         }
 
+        /// <summary>
+        /// tests that the rotation theme with "Relative Rotation" custom property keeps the initial rotation of target gameobject
```
```suggestion
        /// Tests that the rotation theme with "Relative Rotation" custom property keeps the initial rotation of target GameObject.
```

julenka commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/VisualThemeTests.cs:
```diff
@@ -145,6 +183,44 @@ public IEnumerator TestScaleTheme()
         }
 
+        /// <summary>
+        /// tests that the scale theme with "Relative Scale" custom property also take into account the initial scale of target gameobject
```
```suggestion
        /// Tests that the scale theme with "Relative Scale" custom property also takes into account the initial scale of target GameObject.
```

julenka commented on Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs:
```diff
@@ -29,6 +30,23 @@ private static void ReserializeMaterials()
+        {
+            Object[] selectedAssets = Selection.GetFiltered(typeof(Object), SelectionMode.DeepAssets);
+
+            //transform asset object to asset paths
```
```suggestion
            // Transform asset object to asset paths.
```

keveleigh commented:
/azp run

julenka commented on Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs:
```diff
@@ -29,6 +30,23 @@ private static void ReserializeMaterials()
         }
 
+        [MenuItem("Assets/Mixed Reality Toolkit/Reserialize Selection")]
+        public static void ReserializeSelection()
```
@gilbdev could you please add comments for this public method?

julenka replied:
Ah just realized that I merged a bit early. No worries on these comments, if you have time and could please add these in another PR (you can reference the same issue) that would be great.

gilbdev replied:
Sadly i commit this after on but forgot to push 😝

julenka commented:
Thank you for your contribution!
